### ISSUE TITLE
Panics when searching for Roku devices

### ISSUE THREAD

**Issue by tblair** [wbrickner/mote#1]
I get the following panic message when running cargo-mote: (RUST_BACKTRACE=full)

> thread 'tokio-runtime-worker' panicked at 'Failed to parse device info from response: Custom { field: "missing field `network-name`" }', /Users/****/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-mote-0.1.0/src/devices/discovery.rs:90:
  stack backtrace:
    0:        0x105f5f956 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha6f5ddd5259d3e16
    1:        0x105ed9e4d - core::fmt::write::hefe95a44532fe6ed
    2:        0x105f5f206 - std::io::Write::write_fmt::hf513c99961415bff
    3:        0x105f5eb99 - std::panicking::default_hook::{{closure}}::h70a3a940826edea5
    4:        0x105f5e2a9 - std::panicking::rust_panic_with_hook::h102f8bee4e0ef4c7
    5:        0x105f81afa - std::panicking::begin_panic_handler::{{closure}}::hb72eee9aad2e147c
    6:        0x105f81a68 - std::sys_common::backtrace::__rust_end_short_backtrace::h372ff87ecb2667f3
    7:        0x105f81a10 - _rust_begin_unwind
    8:        0x105fc81ef - core::panicking::panic_fmt::h261fd45d36f74dfa
    9:        0x105fc87e5 - core::option::expect_none_failed::hf2a36714418ff407
   10:        0x105eab756 - cargo_mote::devices::discovery::Discoverer::lookup_roku_device_info::{{closure}}::h16454b9071ae71d4
   11:        0x105e9a031 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hcea565d9e93e5275
   12:        0x105e69767 - tokio::runtime::task::raw::poll::ha48ec0bc49417ebc
   13:        0x105f97d79 - tokio::runtime::thread_pool::worker::Context::run_task::hcdfddbb7b7e961bd
   14:        0x105f94eac - tokio::runtime::task::raw::poll::h064775fdca5cdce5
   15:        0x105f9414b - std::sys_common::backtrace::__rust_begin_short_backtrace::h66ec1826f419c367
   16:        0x105f93ccf - core::ops::function::FnOnce::call_once{{vtable.shim}}::hdaeed32a3c6228dc
   17:        0x105f897fd - std::sys::unix::thread::Thread::new::thread_start::h93dd3097fa4fa219
   18:     0x7fff20374950 - __pthread_start

I have 3 Rokus on my network, but I believe the problematic one is connected via ethernet, so it has no "network-name" field to indicate the WiFi network.

Per your suggestion, `GET http://{ip}:8060/query/device-info` to that Roku yields:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<device-info>
	<udn>29680004-5808-10c5-80ce-d8313437c41e</udn>
	<serial-number>************</serial-number>
	<device-id>************</device-id>
	<advertising-id>********-****-****-****-************</advertising-id>
	<vendor-name>Roku</vendor-name>
	<model-name>Roku Ultra</model-name>
	<model-number>4660X</model-number>
	<model-region>US</model-region>
	<is-tv>false</is-tv>
	<is-stick>false</is-stick>
	<ui-resolution>1080p</ui-resolution>
	<supports-ethernet>true</supports-ethernet>
	<wifi-mac>d8:31:34:37:c4:1e</wifi-mac>
	<wifi-driver>realtek</wifi-driver>
	<has-wifi-extender>false</has-wifi-extender>
	<has-wifi-5G-support>true</has-wifi-5G-support>
	<can-use-wifi-extender>true</can-use-wifi-extender>
	<ethernet-mac>d8:31:34:37:c4:1f</ethernet-mac>
	<network-type>ethernet</network-type>
	<friendly-device-name>Bedroom Roku</friendly-device-name>
	<friendly-model-name>Roku Ultra</friendly-model-name>
	<default-device-name>Roku Ultra - YJ004F574926</default-device-name>
	<user-device-name>Bedroom Roku</user-device-name>
	<user-device-location>Bedroom</user-device-location>
	<build-number>469.40E04200A</build-number>
	<software-version>9.4.0</software-version>
	<software-build>4200</software-build>
	<secure-device>true</secure-device>
	<language>en</language>
	<country>US</country>
	<locale>en_US</locale>
	<time-zone-auto>true</time-zone-auto>
	<time-zone>US/Pacific</time-zone>
	<time-zone-name>United States/Pacific</time-zone-name>
	<time-zone-tz>America/Los_Angeles</time-zone-tz>
	<time-zone-offset>-480</time-zone-offset>
	<clock-format>12-hour</clock-format>
	<uptime>759929</uptime>
	<power-mode>PowerOn</power-mode>
	<supports-suspend>false</supports-suspend>
	<supports-find-remote>true</supports-find-remote>
	<find-remote-is-possible>true</find-remote-is-possible>
	<supports-audio-guide>true</supports-audio-guide>
	<supports-rva>true</supports-rva>
	<developer-enabled>false</developer-enabled>
	<keyed-developer-id/>
	<search-enabled>true</search-enabled>
	<search-channels-enabled>true</search-channels-enabled>
	<voice-search-enabled>true</voice-search-enabled>
	<notifications-enabled>true</notifications-enabled>
	<notifications-first-use>false</notifications-first-use>
	<supports-private-listening>true</supports-private-listening>
	<headphones-connected>false</headphones-connected>
	<supports-ecs-textedit>true</supports-ecs-textedit>
	<supports-ecs-microphone>true</supports-ecs-microphone>
	<supports-wake-on-wlan>false</supports-wake-on-wlan>
	<has-play-on-roku>true</has-play-on-roku>
	<has-mobile-screensaver>true</has-mobile-screensaver>
	<support-url>roku.com/support</support-url>
	<grandcentral-version>4.2.38</grandcentral-version>
	<trc-version>3.0</trc-version>
	<trc-channel-version>2.9.42</trc-channel-version>
	<davinci-version>2.8.20</davinci-version>
</device-info>
```

**Comment by wbrickner** [comment]
Funny, I don't own any ethernet Roku devices so I never encountered this.
Good to know that SSDP is working correctly across the two link types though.

I've applied a patch, would you try out the `ethernet-support` branch?

`cargo install --git https://github.com/wbrickner/mote/tree/ethernet-support`

should do the trick.

**Comment by tblair** [comment]
That branch mostly fixed it, but it still ran into an error. For some reason, it's trying to make a device-info request to all 3 Rokus *and my router*. The 3 Rokus respond fine, but the router obviously sends back a 404 which causes it to error out.

Applying the following patch seems to fix it, though you may want to solve the problem differently:

```diff
diff --git a/src/devices/discovery.rs b/src/devices/discovery.rs
index b545709..3351ae1 100644
--- a/src/devices/discovery.rs
+++ b/src/devices/discovery.rs
@@ -48,7 +48,8 @@ impl Discoverer {
         spawn(async move {
           join_all(lookup_futures)
             .await
-            .iter()
+            .into_iter()
+            .filter_map(std::convert::identity)
             .for_each(|result|
               tx
                 .unbounded_send(result.clone())
@@ -67,7 +68,7 @@ impl Discoverer {
       }
     });
   }
-  async fn lookup_roku_device_info(mut location: std::net::SocketAddr) -> Device {
+  async fn lookup_roku_device_info(mut location: std::net::SocketAddr) -> Option<Device> {
     let request = Client::new().request(
         Request::builder()
           .method(Method::GET)
@@ -77,8 +78,11 @@ impl Discoverer {
           .body(Body::empty())
           .expect("Failed to construct request when investigating Roku device info")
       )
-      .await
-      .expect("Failed to get Roku device info");
+      .await;
+    let request = match request {
+      Err(_) => return None,
+      Ok(r) => r,
+    };
 
     // Unfortunately there is no way to parse directly from bytes so
     // we must use a conversion to get our hands on a (verified) utf8 str
@@ -119,10 +123,10 @@ impl Discoverer {
     location.set_port(8060);
 
     // intermediate struct => general device struct
-    Device {
+    Some(Device {
       variant: DeviceType::Roku,
       location,
       info
-    }
+    })
   }
 }
\ No newline at end of file
```

**Comment by wbrickner** [comment]
Fixed by 015d611, plus independent async lookup of device details for lower latency.
